### ISSUE TITLE
Connections to ArangoDB dropping under load

### ISSUE THREAD

**Issue by Calrain** [arangodb/arangodb#7335]
## My Environment

* __ArangoDB Version__:        3.3.19
* __Storage Engine__:             RocksDB
* __Deployment Mode__:       Single Server
* __Deployment Strategy__:   Running as Windows Service
* __Configuration__:               Single Server
* __Infrastructure__:               Own VM
* __Operating System__:        Windows 2016 Server - with GUI
* __Used Package__:              ArangoDB provided installer for Windows

## Component, Query & Data

__Affected feature__:
Resilience testing of Node.js script loading data into ArangoDB purely by Foxx REST endpoints and after a while we get internal errors on ArangoDB followed by ECONNREFUSED and ECONNRESET errors.

Resolution is only to restart the ArangoDB process.

No signs of memory leakage in the application.

The Node application communicates with ArangoDB via REST API endpoints only, and those endpoints perform a lot of actions on the underlying database, such as create collections, load thousands of documents, create thousands of edges, create indexes, queue Foxx jobs, and in the background Foxx is processing many jobs from the standard internal Foxx queue.

If we run this process once, it takes about 2 minutes, and the server runs fine.

If we run this process 100 times in a row, we usually have the ArangoDB process crash in a consistent manner which usually looks like:
1. An error will occur from a Foxx route, such as 'cannot create index due to conflict' on a collection which has no documents in it, and the index can be created manually just fine.
2. 400 Internal Server Errors start coming from Foxx 
3. The Node.js job keeps hitting the Foxx endpoint and then Foxx responds with:
    connect ECONNREFUSED xxx.xxx.xxx.xxx:8530\",\"myendpoint\":\"Error: connect ECONNREFUSED xxx.xxx.xxx.xxx:8530\\n    at Object._errnoException (util.js:1022:11)\\n    at _exceptionWithHostPort (util.js:1044:20)\\n    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1182:14)
4. Then the Node.js jobs gets errors like:
Error: read ECONNRESET\\n    at _errnoException (util.js:1022:11)\\n    at TLSWrap.onread (net.js:615:25)

We do need to improve error handling on the Node.js job to stop hitting Foxx when these errors come back, but the issue is why is Foxx failing in code that works for 50 consecutive calls, but then will fail somewhere between 51 and 100 calls.

Each cycle of Foxx calls (repeated up to 100 times) will create three new collections in the database, add indexes to those collections, load thousands of documents and edges, queue Foxx jobs which immediately run (threads = 0, v8-contexts = 0, 4 core CPU), and update many documents within those three collections. Then the next cycle repeats that with 3 new collections, and then manages those collections in exactly the same fashion.

There aren't issues with RAM, Disk, CPU on the machine, and once our Node.js script finally crashes out the only way to get ArangoDB responsive is to restart the service.

We are using a Self Signed SSL certificate on port 8530.

__Problem__:
It's difficult to share the code due to the complexity of the solution, but I was hoping that the ArangoDB devs may understand why a Foxx application could fail like this.

It's not clear where the error starts from, but it seems to be something at a lower level, because we see issues like Indexes failing to be created on collections that contain no documents, due to a 'conflict', even though that exact process has occurred on 50 other collections in exactly the same way, just minutes before.

Once that error happens, we then start to see ECONNRESET and ECONNREFUSED errors happening with existing calls into Foxx.

We don't have a high number of connections into Foxx, as the Node.js job is reusing the same connection the whole time.

__Expected result__:
ArangoDB to maintain and manage connections even under heavy sustained load.

Does anyone have ideas as to why this may be happening, or pointers I can take to look at what to investigate next?

**Comment by graetzer** [comment]
Dropping / creating collections under load is probably the problem. Can you reproduce these issues on linux and produce a cordump once you see the  _ECONNRESET_ and _ECONNREFUSED_ errors ?

**Comment by mpoeter** [comment]
@graetzer Do you keep the PDB files of your windows builds (e.g., using a symbol server)? With the according PDB files available I could have a look at a windows minidump.

**Comment by dothebart** [comment]
@baslr yes we do! https://github.com/arangodb/arangodb/blob/devel/README_maintainers.md#windows-debugging - need to fix that - its now `download.arangodb.com/repositories/symsrv[xx]/`

**Comment by Calrain** [comment]
Thank you, I will try and get this done tomorrow on the Windows server, thanks for the link to the tool.

**Comment by Calrain** [comment]
Thanks for the info on procdump, we ran that over the arangod process and it never crashed, rather our node.js application was crashing. We didn't see the information on the crash until we also ran procdump on the node application.

It looks like the refused connections into the arango process were coming from malformed calls from Node because it ran out of memory.

Thanks for the heads up on this, I hope other people can use this technique to help with troubleshooting connectivity issues on Windows.

I'll close the ticket as it is an issue now on our code's end.

Thanks!
